### ISSUE TITLE
Crafting URLs with data to an add-on with ingress enabled

### ISSUE THREAD

**Issue by sabeechen** [home-assistant/hassio#1044]
**Home Assistant release with the issue:**
0.91.3

**Operating environment (HassOS/Generic):**
HassOS 2.11

**Supervisor logs:**
Not relevant

**Description of problem:**
I've written an [add-on](https://github.com/sabeechen/hassio-google-drive-backup) that backs up snashots to Google Drive.  It exposes a web ui for looking at the backup status, and naturally I'd like it to to support the new ingress feature.  The problem is that I depend on persisting an OAuth  token from Google through redirects, and I can't figure out how to get this to work with ingress.

Getting oauth credentials form Google goes something  like this:
* **Step 1:** The user clicks a link in my add-on that sends them to an external domain I maintain.
* **Step 2:** The domain redirects to Google authentication endpoint, where they log in and give permisson for the add-on to access their account.
* **Step 3:** Google redirects them back to my domain with a secret, which I'm able to  turn into a long-lived token granting restricted access to the user's Google Drive.
* **Step 4:** My domain redirects the user back to the add-on web UI (which it knows about from step 1) and passes the token for future use as a GET parameter to an endpoint exposed by the add-on.

The issue is with step 4.  Without ingress I can just redirect the user to something like http://hassio.local:1627/savetoken?token=secret_value and the add-on's webserver is able to see and persist the GET value ("token=secret_value") for future use.  With ingress enabled I've tried to get access to url GET parameters but have had no success.  I'm a little fuzzy about how ingress **actually** works beyond putting things in an iframe.  I've tried:
* Adding get paramaters to the ingress-ed web UI Address, eg by redirecting in step 4 to something like https://hassio.local:8123/hassio/ingress/slug/savetoken?token=secret_value, but I just get a pop-up at that address saying `Failed to fetch add-on info`
* After playing with the links in ingress pages for a while, I saw that it often creates links with a format like https://hassio.local:8123/api/hassio_ingress/long_seemingly_random_string/some/sub/path, however trying to redirect to a sub-path I know my add-on is serving (eg https://hassio.local:8123/api/hassio_ingress/OX_iTy9yQdGEwwTnQPVfc4pj_qeV-c677yTxUE8k1oU/savetoken?token=secvet_value for this example) returns `HTTP 502 Bad Gateway` and I haven't been able to reverse engineer it any further.

So to sum up, is there a way to programatically pass information from the URL to an ingress-ed add-on's server?  Right now I'm trying to direct the users back to the add-on's web UI with information in the URL it needs to work, but I'd be happy with any mechanism that lets me pass information to the add-on from a redirect.

@pvizeli Since I can see you've been doing a lot of the work for ingress support

**Comment by pvizeli** [comment]
The URL `https://hassio.local:8123/api/hassio_ingress/OX_iTy9yQdGEwwTnQPVfc4pj_qeV-c677yTxUE8k1oU/savetoken?token=secvet_value` is correct. But it's important that the session cookies is still valid and attached to this call. But without valid cookie, you would receive a 401.

Do you see any kind of error inside log? (use debug on home assistant).
We log if the 502 is from us and not from your add-on.

**Comment by sabeechen** [comment]
The users browser is the one hitting the enpoint, so they should have the right cookies.
When I hit the endpoint, the supervisor logs say:
> 19-04-17 08:10:01 ERROR (MainThread) [hassio.api.ingress] Ingress error: Cannot connect to host hassio.local:8123 ssl:None [Name does not resolve]

But when I go to a non existant sub-dir, I get a 404 from my webserver (which is correct) so I know the requests are going through.  Now that I'm looking at it fresh, I think the issue is that I send a redirect on the "savetoken" URL, and thats causing the problem.  Thats something I can work around, I can just do a javascript redirect instead.

 How does an add-on get the "OX_iTy9yQdGEwwTnQPVfc4pj_qeV-c677yTxUE8k1oU" string?

**Comment by pvizeli** [comment]
I.e. with bashio: `bashio::addon.ingress_entry`

**Comment by pvizeli** [comment]
https://github.com/home-assistant/hassio/blob/dev/hassio/api/ingress.py#L136-L138

I can't see why he should not forward it to... Maybe you have an idea?

**Comment by pvizeli** [comment]
![image](https://user-images.githubusercontent.com/15338540/56272864-e960be00-60fb-11e9-814f-dec569d7a62f.png)

Looks fine on my test system. It load the error from nginx on add-on

**Comment by sabeechen** [comment]
I think the problem I ran into is that my add-on's webserver send a an HTTP redirect to "/" at that address, and that redirect doesn't get translated through the ingress endpoint and somehow comes back as a 502, see screenshot:
![image](https://user-images.githubusercontent.com/21026774/56304084-90dcf100-60fa-11e9-8135-38c3132e1005.png)

I'm not sure if the ingress endpoint **should** pass through header-based redirects, and I can work around it by serving a page there that does a plain-old javascript redirect instead, so I'd consider this resolved.

Thanks for the reference to bashio, it looks like I can get the ingress entrypoint from http://hassio/addons/self/info (I'm using python)
I think that gives me enough to get ingress working well for my add-on.  Thanks for your help and the fast responses!
